Intro to graph ML blog post (#756)
huggingface/blog: intro-graphml.md

---
title: "Introduction to Graph Machine Learning" 
thumbnail: /blog/assets/101_decision-transformers-train/thumbnail.gif
---

# Introduction to Graph Machine Learning

<div class="blog-metadata">
    <small>Published December 23, 2022.</small>
    <a target="_blank" class="btn no-underline text-sm mb-5 font-sans" href="https://github.com/huggingface/blog/blob/main/intro-graphml.md">
        Update on GitHub
    </a>
</div>

<div class="author-card">
    <a href="/clefourrier"> 
        <img class="avatar avatar-user" src="https://aeiljuispo.cloudimg.io/v7/https://s3.amazonaws.com/moonup/production/uploads/1644340617257-noauth.png?w=200&h=200&f=face" title="Gravatar">
        <div class="bfc">
            <code>clefourrier</code>
            <span class="fullname">Clémentine Fourrier</span>
        </div>
    </a>
</div>

## Graphs

### What is a graph?

In its essence, a graph is a description of items linked by relations.

> Examples of graphs include social networks (Twitter, Mastodon, any citation networks linking papers and authors), molecules, knowledge graphs (such as UML diagrams, encyclopedias, any website with hyperlinks between its pages), sentences expressed as their syntactic trees, any 3D mesh, and more! It is therefore not hyperbolic to say that graphs are everywhere.

Author: Add an extra line between the "intro sentence" and the following paragraph?

> Examples of graphs include social networks (Twitter, Mastodon, any citation networks linking papers and authors), molecules, knowledge graphs (such as UML diagrams, encyclopedias, any website with hyperlinks between its pages), sentences expressed as their syntactic trees, any 3D mesh, and more! It is therefore not hyperbolic to say that graphs are everywhere.

Reviewer: Yes, that's right.

However, though these representations seem familiar, do not be fooled!
Graphs are very different from typical objects used in ML because their topology is more complex than just "a sequence" (text, audio) or "a ordered grid" (image/video/...): even if they can be represented as lists or matrices, their representation should not be considered an ordered object!

But what does this mean? Well, if you have a sentence, and you shuffle its words, or an image, and you shuffle its columns, you create a new sentence or new image. 

<p float="left">
  <img src="/assets/124_intro-graphml/hf_logo.png" width="100" />
  <img src="/assets/124_intro-graphml/hf_logo_shuffled.png" width="100" /> 
</p>

> *On the right, the Hugging Face - on the left, a shuffled Hugging Face, which is quite a different new image.*

Author: Is there a usual way to format captions in HFs blog posts? (it's supposed to be the caption of the above image)

> *On the right, the Hugging Face - on the left, a shuffled Hugging Face, which is quite a different new image.*

Reviewer: I don't believe so. You can follow an example markdown [here](https://raw.githubusercontent.com/huggingface/blog/main/clipseg-zero-shot.md) to see how the authors have used the `<figure>` tag to put in captions.

  <img src="/assets/124_intro-graphml/graph_adjacency_shuffled.png" width="30%" /> 
</p>

*On the left, a small graph (nodes in yellow, edges in orange). In the center, its adjacency matrix, with columns and row ordered in the alphabetical node order: on the row for node A (first row), we can read that it is connected to E and C. On the right, a shuffled adjacency matrix (the columns are no longer sorted alphabetically), which is also a valid representation of the graph: A is still connected to E and C.*

## Graph representations through ML

The usual process to work on graphs with machine learning is to, first, generate a meaningful representation for your items of interest (which can be nodes, edges, or full graphs depending on your task), then, use these to train a predictor for your studied task. We want (as in other modalities) to constrain the mathematical representations of your objects so that similar objects are mathematically close. However, similarity is hard to define strictly in graph ML: for example, are two nodes more similar when they have the same labels or the same neighbors?

Note: *In the following sections, we will focus on generating node representations. 

> Once you have node-level representations, it is possible to obtain edge or graph level information from it. For edge level information, you can simply concatenate node pair representations, or simply do a dot product. For graph level information, it is possible to do a global pooling (average/sum/...), but it will smooth and loose information over the graph -- a recursive hierarchical pooling can make more sense, or to add a virtual node, connected to all other nodes in the graph, and use its representation as the overall graph representation.*

Author: Exactly! I will rephrase it.

### Pre-neural approaches

#### Simply using engineered features

> Pre-neural networks, graphs and their items of interests could be represented as combination of features, in a task specific fashion. (Now, these features are used for data augmentation and semi-supervised learning.) 

Author: To the general definitions in wikipedia, or to specific papers in graph ML?

> Pre-neural networks, graphs and their items of interests could be represented as combination of features, in a task specific fashion. (Now, these features are used for data augmentation and semi-supervised learning.) 

Reviewer: Sorry for being unclear here and there. Deeply apologize. 

I meant hyperlinks specific to graph ML.

> Pre-neural networks, graphs and their items of interests could be represented as combination of features, in a task specific fashion. (Now, these features are used for data augmentation and semi-supervised learning.) 

Author: Absolutely not a problem, thank you so much for having taken the time to go through this!

> Pre-neural networks, graphs and their items of interests could be represented as combination of features, in a task specific fashion. (Now, these features are used for data augmentation and semi-supervised learning.) 

Author: I'll see if I have specific references that I have read on the topic.

> Node-level features can give information about importance (how important is this node for the graph?) and/or structure based (what is the shape of the graph around the node?), and can be combined.

Author: There are also edge and graph level features, I will add them back (I removed them to focus on node representation for clarity, but exhaustivity is possibly better)

However, these methods have limits: they cannot obtain embeddings for new nodes, do not capture structural similarity between nodes finely, and cannot use added features.

## Graph Neural Networks

Neural networks can (hopefully) generalise to unseen data. What should a good neural network be, to work on graphs, given the representation constraints we evoked earlier?

It should:

- be permutation invariant:
    - Equation: \\(f(P(G))=f(G)\\) with f the network, P the permutation function, G the graph
    - Explanation: the representation of a graph and its permutations should be the same after going through the network)
- be permutation equivariant
    - Equation: \\(P(f(G))=f(P(G))\\) with f the network, P the permutation function, G the graph

>     - Explanation: permuting the nodes before processing should be equivalent to permuting their representations

Author: Would "permuting the nodes before passing them to the network should be equivalent to permuting their representations" be clearer?

>     - Explanation: permuting the nodes before processing should be equivalent to permuting their representations

Reviewer: Yes. Much more explicit and clear.

- [Graph Convolutional Networks](https://tkipf.github.io/graph-convolutional-networks/) averages the normalised representation of the neighbours for a node (most GNNs are actually GCNs);
- [Graph Attention Networks](https://petar-v.com/GAT/) learn to weigh the different neighbours based on their importance (like transformers);
- [GraphSAGE](https://snap.stanford.edu/graphsage/) samples neighbours at different hops before aggregating their information in several steps with max pooling.
- [Graph Isomorphism Networks](https://arxiv.org/pdf/1810.00826v3.pdf) aggregates representation by applying an MLP to the sum of the neighbours node representations.

**Choosing an aggregation**: Some aggregation techniques (notably mean/max pooling) can encounter failure cases when creating representations which finely differentiate nodes with different neighbourhoods of similar nodes (ex: through mean pooling, a neighbourhood with 4 nodes, represented as 1,1,-1,-1, averaged as 0, is not going to be different from one with only 3 nodes represented as -1, 0, 1).

### GNN shape and the over-smoothing problem

At each new layer, the node representation includes more and more nodes. 

> A node through the first layer, is the aggregation of its direct neighbours. Through the second layer, it is still the aggregation of its direct neighbours, but this time, their representations now include their own neighbours (from the first layer)! If your network had too many layers, there is a risk that each node becomes an aggregation of the full graph (and that nodes representations will no longer be different). This is called **the oversmoothing problem**.

Author: Because, after n layers, the representation of all nodes becomes an aggregation of all their neighbours at distance n, therefore, of the full graph, if its diameter is smaller than n - is this explanation better, or should i go step by step with examples?

> A node through the first layer, is the aggregation of its direct neighbours. Through the second layer, it is still the aggregation of its direct neighbours, but this time, their representations now include their own neighbours (from the first layer)! If your network had too many layers, there is a risk that each node becomes an aggregation of the full graph (and that nodes representations will no longer be different). This is called **the oversmoothing problem**.

Reviewer: I think this is better. If there's any visual depiction you can present via a figure, feel free to add it (not a must-have IMO).

This can be solved by :

- scaling the GNN to have a layer number small enough to not approximate each node as the whole network (by first analysing the graph diameter and shape)
- increasing the complexity of the layers
- adding non message passing layers to process the messages (such as simple MLPs)
- adding skip-connections.

The oversmoothing problem is an important area of study in graph ML, as it prevents GNNs to scale up, like Transformers have been shown to in other modalities.

## Graph Transformers

> A Transformer without its positional encoding layer is permutation invariant, and Transformers are known to scale well, so recently, people have started looking at adapting Transformers to graphs ([Survey)](https://github.com/ChandlerBang/awesome-graph-transformer). Most methods focus on the best ways to represent graphs, by looking for the best features and best ways to represent positional information, and changing the attention to fit this new data.

Author: I think it's because no general architecture has been found at the moment for graphs, much like in NLP a decade ago, where neural methods were augmented with rule based/statistical features - do you think this should be mentioned?

> Here are some methods which seemed interesting to me, and which got SOTA or close on one of the hardest available benchmarks as of writing, Stanford's Open Graph Benchmark:

Reviewer: ```suggestion
Here are some interesting methods which got state-of-the-art results or close on one of the hardest available benchmarks as of writing, [Stanford's Open Graph Benchmark](https://ogb.stanford.edu/):
```

- *Graph Transformer for Graph-to-Sequence Learning* (Cai and Lam, 2020) introduced a Graph Encoder, which represents nodes as a concatenation of their embeddings and positional embeddings, node relations as the shortest paths between them, and combine both in a relation-augmented self attention.
- *Rethinking Graph Transformers with Spectral Attention* (Kreuzer et al, 2021) introduced Spectral Attention Networks (SANs). These combine node features with learned positional encoding (computed from Laplacian eigenvectors/values), to use as keys and queries in the attention, with attention values being the edge features.
- *GRPE: Relative Positional Encoding for Graph Transformer* (Park et al, 2021) introduced the Graph Relative Positional Encoding Transformer. It represents a graph by combining a graph-level positional encoding with node information, edge level positional encoding with node information, and combining both in the attention.

> - *Global Self-Attention as a Replacement for Graph Convolution* (Hussain et al, 2021) introduced the Edge Augmented Transformer. This architecture embeds nodes and edges separately, and aggregates them in a modified attention.

Reviewer: Hyperlink each of the papers?

> My favorites are:

Reviewer: Needs rewording. :D

Author: The last sentence? ^^

Reviewer: Yeah. In the blog post, "My" and similar phrases could be avoided.

  <figcaption>On the right, the Hugging Face logo - on the left, a shuffled Hugging Face, which is quite a different new image.</figcaption>
</figure>

This is not the case for a graph: if you shuffle its edge list, or the columns of its adjacency matrix, it is still the same graph. (We explain this more formally a bit lower, look for permutation invariance).

<figure class="image table text-center m-0 w-full">
  <img src="/assets/124_intro-graphml/graph.png" width="30%" />
  <img src="/assets/124_intro-graphml/graph_adjacency.png" width="30%" /> 
  <img src="/assets/124_intro-graphml/graph_adjacency_shuffled.png" width="30%" /> 
  <figcaption>On the left, a small graph (nodes in yellow, edges in orange). In the center, its adjacency matrix, with columns and row ordered in the alphabetical node order: on the row for node A (first row), we can read that it is connected to E and C. On the right, a shuffled adjacency matrix (the columns are no longer sorted alphabetically), which is also a valid representation of the graph: A is still connected to E and C.</figcaption>
</figure>

## Graph representations through ML

> The usual process to work on graphs with machine learning is to, first, generate a meaningful representation for your items of interest (which can be nodes, edges, or full graphs depending on your task), then, use these to train a predictor for your studied task. We want (as in other modalities) to constrain the mathematical representations of your objects so that similar objects are mathematically close. However, similarity is hard to define strictly in graph ML: for example, are two nodes more similar when they have the same labels or the same neighbours?

Reviewer 2: ```suggestion
The usual process to work on graphs with machine learning is first to generate a meaningful representation for your items of interest (nodes, edges, or full graphs, depending on your task), then use these to train a predictor for your target task. We want (as in other modalities) to constrain the mathematical representations of your objects so that similar objects are mathematically close. However, the similarity is hard to define strictly in graph ML: for example, are two nodes more similar when they have the same labels or the same neighbors?
```

Note: *In the following sections, we will focus on generating node representations. 

> Once you have node-level representations, it is possible to obtain edge or graph-level information from it. For edge-level information, you can simply concatenate node pair representations or simply do a dot product. For graph-level information, it is possible to do a global pooling (average/sum/...) on the concatenated tensor of all the node-level representations. Still, it will smooth and lose information over the graph -- a recursive hierarchical pooling can make more sense, or to add a virtual node, connected to all other nodes in the graph, and use its representation as the overall graph representation.*

Reviewer 2: ```suggestion
Once you have node-level representations, it is possible to obtain edge or graph-level information. For edge-level information, you can concatenate node pair representations or do a dot product. For graph-level information, it is possible to do a global pooling (average, sum, etc.) on the concatenated tensor of all the node-level representations. Still, it will smooth and lose information over the graph -- a recursive hierarchical pooling can make more sense, or add a virtual node, connected to all other nodes in the graph, and use its representation as the overall graph representation.*
```

### Pre-neural approaches

#### Simply using engineered features

> Pre-neural networks, graphs and their items of interests could be represented as combination of features, in a task specific fashion. Now, these features are still used for data augmentation and [semi-supervised learning](https://arxiv.org/abs/2202.08871), though [more complex feature generation methods](https://arxiv.org/abs/2208.11973) exist; it can be important to find how best to provide them to your network depending on your task. 

Reviewer 2: ```suggestion
Pre-neural networks, graphs, and their items of interest could be represented as a combination of features in a task-specific fashion. Now, these features are still used for data augmentation and [semi-supervised learning](https://arxiv.org/abs/2202.08871), though [more complex feature generation methods](https://arxiv.org/abs/2208.11973) exist; they can be essential to find how best to provide them to your network depending on your task. 
```

**Node-level** features can give information about importance (how important is this node for the graph?) and/or structure based (what is the shape of the graph around the node?), and can be combined.

> The node **centrality** measures the node importance in the graph. It can be computed recursively by summing the centrality of each node’s neighbours until convergence, or through shortest distance measures between nodes, for example. The node **degree** is the quantity of direct neighbours it has. The **clustering coefficient** measures how connected the node neighbours are. **Graphlets degree vectors** count how many different graphlets are rooted at a given node, where graphlets are all the mini graphs you can create with a given number of connected nodes (with 3 connected nodes, you can have a line with 2 edges, or a triangle with 3 edges).

Reviewer 2: ```suggestion
The node **centrality** measures the node importance in the graph. It can be computed recursively by summing the centrality of each node’s neighbors until convergence, or through shortest distance measures between nodes, for example. The node **degree** is the number of direct neighbors it has. The **clustering coefficient** measures how connected the node neighbors are. **Graphlets degree vectors** count how many different graphlets are rooted at a given node, where graphlets are all the mini graphs you can create with a given number of connected nodes (with three connected nodes, you can have a line with two edges or a triangle with three edges).
```

![Graphlets list for 2 to 5 connected nodes](assets/124_intro-graphml/graphlets.png)

*The 2-to 5-node graphlets (Pržulj, 2007)*

**Edge-level** features complement the representation with more detailed information about the connectedness of the nodes, and include the **shortest distance** between two nodes, their **common neighbours**, and their **Katz index** (which is the number of possible walks of up to a certain length between two nodes - it can be computed directly from the adjacency matrix).

> **Graph level features** contain high-level information about graph similiarity and specificities. Total **graphlet counts**, though computationally expensive, provide information about the shape of sub-graphs. **Kernel methods** measure similarity between graphs through different "bag of nodes" method (similar to bag of words).

Reviewer 2: ```suggestion
**Graph level features** contain high-level information about graph similarity and specificities. Total **graphlet counts**, though computationally expensive, provide information about the shape of sub-graphs. **Kernel methods** measure similarity between graphs through different "bag of nodes" methods (similar to bag of words).
```

### Walk-based approaches

> [**Walk-based approaches**](https://en.wikipedia.org/wiki/Random_walk) use the probability of visiting a node j from a node i on a random walk to define similarity metrics; these approaches combine both local and global information. [**Node2Vec**](https://snap.stanford.edu/node2vec/), for example, simulates random walks between nodes of a graph, then processes these walks with a skip-gram, [much like we would do with words in sentences](https://arxiv.org/abs/1301.3781), to compute embeddings. These approaches can also be used to [accelerate computations](https://arxiv.org/abs/1208.3071) of the [**Page Rank method**](http://infolab.stanford.edu/pub/papers/google.pdf), which assign an importance score to each node (based on its connectivity to other nodes, evaluated as its frequency of visit by random walk, for example).

Reviewer 2: ```suggestion
[**Walk-based approaches**](https://en.wikipedia.org/wiki/Random_walk) use the probability of visiting a node j from a node i on a random walk to define similarity metrics; these approaches combine both local and global information. [**Node2Vec**](https://snap.stanford.edu/node2vec/), for example, simulates random walks between nodes of a graph, then processes these walks with a skip-gram, [much like we would do with words in sentences](https://arxiv.org/abs/1301.3781), to compute embeddings. These approaches can also be used to [accelerate computations](https://arxiv.org/abs/1208.3071) of the [**Page Rank method**](http://infolab.stanford.edu/pub/papers/google.pdf), which assigns an importance score to each node (based on its connectivity to other nodes, evaluated as its frequency of visit by random walk, for example).
```

However, these methods have limits: they cannot obtain embeddings for new nodes, do not capture structural similarity between nodes finely, and cannot use added features.

## Graph Neural Networks

> Neural networks can (hopefully) generalise to unseen data. What should a good neural network be, to work on graphs, given the representation constraints we evoked earlier?

Reviewer 2: ```suggestion
Neural networks can generalize to unseen data. Given the representation constraints we evoked earlier, what should a good neural network be to work on graphs?
```

It should:

- be permutation invariant:
    - Equation: \\(f(P(G))=f(G)\\) with f the network, P the permutation function, G the graph
    - Explanation: the representation of a graph and its permutations should be the same after going through the network)
- be permutation equivariant
    - Equation: \\(P(f(G))=f(P(G))\\) with f the network, P the permutation function, G the graph
    - Explanation: permuting the nodes before passing them to the network should be equivalent to permuting their representations

Typical neural networks, such as RNNs or CNNs are not permutation invariant. A new architecture, the [Graph Neural Network](https://ieeexplore.ieee.org/abstract/document/1517930), was therefore introduced (initially as a state-based machine).

> A GNN is made of successive layers. A GNN layer represents a node as the combination (**aggregation**) of the representations of its neighbours and itself from the previous layer (**message passing**), plus usually an activation to add some nonlinearity.

Reviewer 2: ```suggestion
A GNN is made of successive layers. A GNN layer represents a node as the combination (**aggregation**) of the representations of its neighbors and itself from the previous layer (**message passing**), plus usually an activation to add some nonlinearity.
```

> **Comparison to other models**: A CNN can be seen as a GNN with fixed neighbour sizes (through the sliding window) and ordering (it is not permutation equivariant). A [Transformer](https://arxiv.org/abs/1706.03762v3) without positional embeddings can be seen as a GNN on a fully-connected input graph.

Reviewer 2: ```suggestion
**Comparison to other models**: A CNN can be seen as a GNN with fixed neighbor sizes (through the sliding window) and ordering (it is not permutation equivariant). A [Transformer](https://arxiv.org/abs/1706.03762v3) without positional embeddings can be seen as a GNN on a fully-connected input graph.
```

### Aggregation and message passing

> There are many ways to aggregate messages from neighbour nodes, summing, averaging, for example. Some notable works following this idea include:

Reviewer 2: ```suggestion
There are many ways to aggregate messages from neighbor nodes, summing, averaging, for example. Some notable works following this idea include:
```

> - [Graph Convolutional Networks](https://tkipf.github.io/graph-convolutional-networks/) averages the normalised representation of the neighbours for a node (most GNNs are actually GCNs);
> - [Graph Attention Networks](https://petar-v.com/GAT/) learn to weigh the different neighbours based on their importance (like transformers);
> - [GraphSAGE](https://snap.stanford.edu/graphsage/) samples neighbours at different hops before aggregating their information in several steps with max pooling.
> - [Graph Isomorphism Networks](https://arxiv.org/pdf/1810.00826v3.pdf) aggregates representation by applying an MLP to the sum of the neighbours node representations.

Reviewer 2: ```suggestion
- [Graph Convolutional Networks](https://tkipf.github.io/graph-convolutional-networks/) averages the normalized representation of the neighbors for a node (most GNNs are actually GCNs);
- [Graph Attention Networks](https://petar-v.com/GAT/) learn to weigh the different neighbors based on their importance (like transformers);
- [GraphSAGE](https://snap.stanford.edu/graphsage/) samples neighbors at different hops before aggregating their information in several steps with max pooling.
- [Graph Isomorphism Networks](https://arxiv.org/pdf/1810.00826v3.pdf) aggregates representation by applying an MLP to the sum of the neighbors' node representations.
```

Author: Is there a preference for American English over British English?

**Choosing an aggregation**: Some aggregation techniques (notably mean/max pooling) can encounter failure cases when creating representations which finely differentiate nodes with different neighbourhoods of similar nodes (ex: through mean pooling, a neighbourhood with 4 nodes, represented as 1,1,-1,-1, averaged as 0, is not going to be different from one with only 3 nodes represented as -1, 0, 1).

### GNN shape and the over-smoothing problem

At each new layer, the node representation includes more and more nodes. 

> A node, through the first layer, is the aggregation of its direct neighbours. Through the second layer, it is still the aggregation of its direct neighbours, but this time, their representations now include their own neighbours (from the first layer). After n layers, the representation of all nodes becomes an aggregation of all their neighbours at distance n, therefore, of the full graph, if its diameter is smaller than n! 

Reviewer 2: ```suggestion
A node, through the first layer, is the aggregation of its direct neighbors. Through the second layer, it is still the aggregation of its direct neighbors, but this time, their representations now include their neighbors (from the first layer). After n layers, the representation of all nodes becomes an aggregation of all their neighbors at distance n, therefore, of the full graph if its diameter is smaller than n! 
```

> If your network had too many layers, there is therefore a risk that each node becomes an aggregation of the full graph (and that node representations converge to the same one for all nodes). This is called **the oversmoothing problem**.

Reviewer 2: ```suggestion
If your network has too many layers, there is a risk that each node becomes an aggregation of the full graph (and that node representations converge to the same one for all nodes). This is called **the oversmoothing problem**
```

This can be solved by :

- scaling the GNN to have a layer number small enough to not approximate each node as the whole network (by first analysing the graph diameter and shape)
- increasing the complexity of the layers
- adding non message passing layers to process the messages (such as simple MLPs)
- adding skip-connections.

The oversmoothing problem is an important area of study in graph ML, as it prevents GNNs to scale up, like Transformers have been shown to in other modalities.

## Graph Transformers

> A Transformer without its positional encoding layer is permutation invariant, and Transformers are known to scale well, so recently, people have started looking at adapting Transformers to graphs ([Survey)](https://github.com/ChandlerBang/awesome-graph-transformer). Most methods focus on the best ways to represent graphs, by looking for the best features and best ways to represent positional information, and changing the attention to fit this new data.

Reviewer 2: ```suggestion
A Transformer without its positional encoding layer is permutation invariant, and Transformers are known to scale well, so recently, people have started looking at adapting Transformers to graphs ([Survey)](https://github.com/ChandlerBang/awesome-graph-transformer). Most methods focus on the best ways to represent graphs by looking for the best features and best ways to represent positional information and changing the attention to fit this new data.
```

Using transformers for graphs is still very much a field in its infancy, but it looks promising, as it could alleviate several limitations of GNNs, such as scaling to larger/denser graphs, or increasing model size without oversmoothing.

# Further resources

If you want to delve deeper, you can look at some of these courses:

- Academic format
    - [Stanford's Machine Learning with Graphs](https://web.stanford.edu/class/cs224w/)
    - [McGill's Graph Representation Learning](https://cs.mcgill.ca/~wlh/comp766/)
- Video format
    - [Geometric Deep Learning course](https://www.youtube.com/playlist?list=PLn2-dEmQeTfSLXW8yXP4q_Ii58wFdxb3C)
- Books
    - [Graph Representation Learning*, Hamilton](https://www.cs.mcgill.ca/~wlh/grl_book/)

> Nice librairies to work on graphs are [PyGeometric](https://pytorch-geometric.readthedocs.io/en/latest/) (for graph ML) and [NetworkX](https://networkx.org/) (to manipulate graphs more generally).

Reviewer 2: ```suggestion
Nice libraries to work on graphs are [PyGeometric](https://pytorch-geometric.readthedocs.io/en/latest/) (for graph ML) and [NetworkX](https://networkx.org/) (to manipulate graphs more generally).
```
